### PR TITLE
[Impeller] Implement tile mode for linear gradient and radial gradient

### PR DESCRIPTION
**Pull request by ColdPaleLight**
fix https://github.com/flutter/flutter/issues/108810

gu5D3BG2IC

cOcwnTXeym

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: impeller/geometry/tile_mode.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (bdero) [inline impeller/geometry/tile_mode.h:1]: `geometry` is primarily used for math utilities. Please place this enum in the entity framework. Perhaps in the `Entity` declaration next to the `BlendMode` enum class.

File: impeller/entity/shaders/tile_mode_utils.glsl [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+const float kTileModeClamp = 0;
```
Comment (bdero) [inline impeller/entity/shaders/tile_mode_utils.glsl:5]: Please add a comment nearby indicating that these values must correspond to the order of the items in the `Entity::TileMode` enum class.

**Review by bdero — COMMENTED** [review]
Overall looking good, thanks! Just some comments about where things should live.

File: impeller/geometry/tile_mode.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#pragma once
+
+namespace impeller {
+
+// An enum to define how to repeat, fold, or omit colors outside of the
+// typically defined range of the source of the colors (such as the
+// bounds of an image or the defining geoetry of a gradient).
```
[the remark this answers is not in the document]
Comment (ColdPaleLight) [reply impeller/geometry/tile_mode.h]: Done.

File: impeller/entity/shaders/tile_mode_utils.glsl [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+const float kTileModeRepeat = 1;
+const float kTileModeMirror = 2;
+const float kTileModeDecal = 3;
+
+float GetInterpolantValue(float t, float tile_mode) {
```
[the remark this answers is not in the document]
Comment (ColdPaleLight) [reply impeller/entity/shaders/tile_mode_utils.glsl]: Done.

File: impeller/entity/shaders/tile_mode_utils.glsl [hunk 5]
```diff
@@ -0,0 +1,21 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (ColdPaleLight) [reply impeller/entity/shaders/tile_mode_utils.glsl]: Done.

File: impeller/geometry/tile_mode.h (hunk 1 above)
Comment (ColdPaleLight) [reply impeller/geometry/tile_mode.h]: Done.

File: impeller/entity/shaders/tile_mode_utils.glsl (hunk 2 above)
Comment (ColdPaleLight) [reply impeller/entity/shaders/tile_mode_utils.glsl]: Done.

File: impeller/entity/entity.h [hunk 6]
```diff
@@ -62,6 +62,29 @@ class Entity {
     kLastAdvancedBlendMode = kLuminosity,
   };
 
+  /// An enum to define how to repeat, fold, or omit colors outside of the
+  /// typically defined range of the source of the colors (such as the
+  /// bounds of an image or the defining geoetry of a gradient).
```
Comment (bdero) [inline impeller/entity/entity.h:67]:
```suggestion
  /// bounds of an image or the defining geometry of a gradient).
```

File: impeller/entity/entity.h [hunk 7]
```diff
@@ -62,6 +62,29 @@ class Entity {
...
+    /// bounds.
+    kClamp,
+
+    /// Repeat the shader's image horizontally and vertically (or both along and
+    /// perpendicular to a gradient's geometry).
+    kRepeat,
+
+    /// Repeat the shader's image horizontally and vertically, alternating
+    /// mirror
+    /// images so that adjacent images always seam.
```
Comment (bdero) [inline impeller/entity/entity.h:79]:
```suggestion
    /// Repeat the shader's image horizontally and vertically, seamlessly
    /// alternating mirrored images.
```

**Review by bdero — APPROVED** [review]
lgtm!

**Pull request merged**